datafusion/expr/src/predicate_bounds.rs

```diff
@@ -0,0 +1,1045 @@
+        Ok(match value {
+            ScalarValue::Null => TernarySet::UNKNOWN,
+            ScalarValue::Boolean(b) => match b {
+                Some(true) => TernarySet::TRUE,
+                Some(false) => TernarySet::FALSE,
+                None => TernarySet::UNKNOWN,
+            },
+            _ => {
+                let b = value.cast_to(&DataType::Boolean)?;
+                Self::try_from(&b)?
+            }
+        })
+    }
+}
+
+/// Computes the output interval for the given boolean expression based on statically
+/// available information.
+///
+/// # Arguments
+///
+/// * `predicate` - The boolean expression to analyze
+/// * `is_null` - A callback function that provides additional nullability information for
+///   expressions. When called with an expression, it should return:
+///   - `Some(true)` if the expression is known to evaluate to NULL
+///   - `Some(false)` if the expression is known to NOT evaluate to NULL
+///   - `None` if the nullability cannot be determined
+///
+///   This callback allows the caller to provide context-specific knowledge about expression
+///   nullability that cannot be determined from the schema alone. For example, it can be used
+///   to indicate that a particular column reference is known to be NULL in a specific context,
+///   or that certain expressions will never be NULL based on runtime constraints.
+///
+/// * `input_schema` - Schema information for resolving expression types and nullability
+///
+/// # Return Value
+///
+/// The function returns a [NullableInterval] that describes the possible boolean values the
+/// predicate can evaluate to. The return value will be one of the following:
+///
+/// * `NullableInterval::NotNull { values: Interval::CERTAINLY_TRUE }` - The predicate will
+///   always evaluate to TRUE (never FALSE or NULL)
+///
+/// * `NullableInterval::NotNull { values: Interval::CERTAINLY_FALSE }` - The predicate will
+///   always evaluate to FALSE (never TRUE or NULL)
+///
+/// * `NullableInterval::NotNull { values: Interval::UNCERTAIN }` - The predicate will never
+///   evaluate to NULL, but may be either TRUE or FALSE
+///
+/// * `NullableInterval::Null { datatype: DataType::Boolean }` - The predicate will always
+///   evaluate to NULL (SQL UNKNOWN in three-valued logic)
+///
+/// * `NullableInterval::MaybeNull { values: Interval::CERTAINLY_TRUE }` - The predicate may
+///   evaluate to TRUE or NULL, but never FALSE
+///
+/// * `NullableInterval::MaybeNull { values: Interval::CERTAINLY_FALSE }` - The predicate may
+///   evaluate to FALSE or NULL, but never TRUE
+///
+/// * `NullableInterval::MaybeNull { values: Interval::UNCERTAIN }` - The predicate may
+///   evaluate to any of TRUE, FALSE, or NULL
+///
+pub(super) fn evaluate_bounds(
+    predicate: &Expr,
+    certainly_null_expr: Option<&Expr>,
+    input_schema: &dyn ExprSchema,
+) -> Result<NullableInterval> {
+    let evaluator = PredicateBoundsEvaluator {
```

Comment:
Is your idea that you might be able to use the `NullableInterval` code instead of this (now that you have fixed the bugs?)